app/src/main/java/org/jellyfin/mobile/player/source/JellyfinMediaSource.kt

```diff
@@ -10,13 +11,15 @@ import org.jellyfin.sdk.model.api.PlayMethod
 import org.jellyfin.sdk.model.api.SubtitleDeliveryMethod
 import java.util.UUID
 
+@Serializable(with = JellyfinMediaSourceSerializer::class)
 class JellyfinMediaSource(
     val itemId: UUID,
     val item: BaseItemDto?,
     val sourceInfo: MediaSourceInfo,
     val playSessionId: String,
     val liveStreamId: String?,
     val maxStreamingBitrate: Int?,
+    var isDownload: Boolean = false,
```

Comment:
> `The constructor(itemId: UUID, item: BaseItemDto?, sourceInfo: MediaSourceInfo, playSessionId: String, liveStreamId: String?, maxStreamingBitrate: Int?, isDownload: Boolean, startTimeTicks: Long?, audioStreamIndex: Int?, subtitleStreamIndex: Int?) has too many parameters. The current threshold is set to 10. [LongParameterList]`

A potential solution for this might be to serialise the params in a DTO to group some of our client side media source state together? Something a bit like this:

```kotlin
data class MediaSourceDetails(
    val itemId: UUID,
    val item: BaseItemDto?,
    val sourceInfo: MediaSourceInfo,
    val playSessionId: String,
    val liveStreamId: String?
)

class JellyfinMediaSource(
    val details: MediaSourceDetails,
    val maxStreamingBitrate: Int?,
    var isDownload: Boolean = false,
    private var startTimeTicks: Long? = null,
    audioStreamIndex: Int? = null,
    subtitleStreamIndex: Int? = null
) 
```

Reply:
That would work! :) 
My only thoughts on that is _what if the users network connection is unstable_... I'm thinking a perfect use case for this PR would be for travelling with a potentially intermittent connection &mdash; an API request might succeed but the user would (should 😁 ) be playing from the local file.

I have yet to do a deep dive in the codebase but what's your thoughts on always defaulting to the local file if it exists in the `DownloadDao` as a viable alternative? We could simply poll the Dao in the `isDownload` predicate?

Reply:
Agree on having the app be useable when the network is down being an important use case.

The most obvious example is being on a plane and downloading a video before boarding, after which one may not have network access.

Reply:
Absolutely! I believe @7ritn's solution of polling the server when the user plays the content to see if there's a connection would work for when there is absolutely no connection (i.e., on a plane).
My concern is _what if the connection is slow or intermittent?_ For example, if you're on a train, the connection might work (if you're lucky!) but, the moment you go in a tunnel, the player will still be following the "I'm not playing local content" behaviour (as the connection was available at the time of polling).

We could also consider an "offline mode" toggle switch that could force use of the `DownloadDao` but implementations in other apps have always been clunky (and still seems excessive for a code quality rule :/).

Reply:
Due to the caching / downloading being handled by exoplayer, if there is a download matching the requested content id available, exoplayer will automatically playback the download. Thus with intermittent or even good connection the download will be preferred. This currently only works if the original quality is used, I'm not sure what the behaviour is if the user chooses a different quality i.e. transcoded stream.

Reply:
If no elegant solution exists, it's also ok to simply suppress that warning, btw. Don't feel pressured by warnings like these. They are a good practice, but not always mandatory.

Reply:
> If no elegant solution exists, it's also ok to simply suppress that warning, btw. Don't feel pressured by warnings like these. They are a good practice, but not always mandatory.

Well we are now working on a solution using inheritance to have RemoteJellyfinMediaSource and a LocalJellyfinMediaSource which also additionally contains download url and oter downloads specs, simplifying the database

Reply:
Done